### ISSUE TITLE
Premium features for donation and  F-droid releases

### ISSUE THREAD

**Issue by 154pinkchairs** [kawaiiDango/pScrobbler#229]
Would be nice if there was a way to access premium features without gApps, since Aurora Store unfortunately doesn't support purchasing new apps so I have to use a secondary smartphone or my PC, whilst I'm also a bit hesitant to give Google my CC data. The simplest way I can think of achieving it is if there was a possibility to get an unique key available after donating via either PayPal, bank transfer, or preferably crypto, such as Monero, where the user would e-mail you TX ID to get that key.

Oh, and since the app is licensed under GPL, it could be published on F-droid, possibly there can be a dedicated package without Google Crashlytics and this way of obtaining premium features as the only one available. Instead, since people who use F-Droid are mostly techies, stack trace could be generated and the app would suggest sending an e-mail with it. It may also be necessary to try replacing ACRCloud with (AcoustID)[https://acoustid.org/]. If you don't have the time to create separate releases for F-droid, I can try becoming a maintainer.

**Comment by Efreak** [comment]
Regarding fdroid, see #2 and #216. I just added pScrobbler to [my automated fdroid repo](https://github.com/efreak/auto-daily-fdroid), which will pull new github releases daily. However, I believe @kawaiiDango includes nonfree libs in the releases, so up to you if you want to install. Anything to one of those issues, the app does work with microg (I use Gapps, so I can't tell you if it works without).

**Comment by 154pinkchairs** [comment]
I also use microG. The thing is it connects to Crashlytics servers, though they can be blocked with Nebulo, pihole, etc. And that other library for scrobbling from mic, which has a libre alternative. Perhaps in July I'll have the time to make and test a more libre build and share the results here.

**Comment by lostb053** [comment]
So is there no other way for me to get premium features?

**Comment by freshgum-bubbles** [comment]
Yeah, I'm agreeable to this. If there was a way to support the app without Google Play, that'd be a lot better.

I'm not sure what F-Droid's terms on pay-gates are (cc @IzzySoft), but -- as a very regular user of Pano -- with the clear amount of effort put into this app I'd be *very* happy to pay a fair amount for it ;-)

It goes without saying that the alternative method wouldn't be included in the Play Store builds, though -- against their policies, from what I understand.

**Comment by kawaiiDango** [comment]
So I made a proof of concept JWT based key generator, mailer and a license checking server, just out of curiosity. It is far from complete as of now. But in the future, I might provide a way for ko-fi donators who have donated above a certain amount, to get a key to unlock premium features for the app downloaded from GitHub or the custom f-droid repo (which in turn downloads from GitHub releases).

**Comment by IzzySoft** [comment]
@freshgum-bubbles apologies, missed your ping back then – but received one for the last comment… So: the suggested approach should be fine – as long as the entire app is open-source. F/LOSS does not mean "must be free as in free beer", but "must be libre" – see [The Four Essential Freedoms of Free Software](https://en.wikipedia.org/wiki/The_Free_Software_Definition#The_Four_Essential_Freedoms_of_Free_Software).

**Comment by kawaiiDango** [comment]
IzzySoft, refer to this comment, https://github.com/kawaiiDango/pano-scrobbler/issues/2#issuecomment-1099163793

If I were to create a non play store build, all of those non free dependencies wont make sense anymore and will be removed, except Acrcloud (since it provides a very useful feature).

But this repo is not buildable as it is. The files containing API keys and other secrets have been gitignored. This will be an issue for Izzy's automated builds.

**Comment by IzzySoft** [comment]
> The files containing API keys and other secrets have been gitignored. This will be an issue for Izzy's automated builds.

Yupp, they'd be. But there are options to it. First let me state two "existing facts":

* IoD takes the APKs from the developers as shipped at their releases. Our "automated builds" are for confirming they correspond to the source they claim to be built from, see: [Reproducible Builds](https://reproducible-builds.org/) and, for some details, [Reproducible Builds, special client support and more in our repo](https://android.izzysoft.de/articles/named/iod-rbs-mirrors-clients) – and find e.g. [my builder here](https://codeberg.org/IzzyOnDroid/rbtlog).
* API keys can always be extracted from the APK itself. They're just not "as obvious in plain sight".

Building on these two facts: you could "camouflage" the keys, e.g. encoding the keys (or their entire config files) using base64 – and then decode that at build-time to create those files/keys. That way they'd still be "out of plain sight", but all builders could handle it easily. Needless to say this does NOT go for the signing keys. Those are only needed by you: our verification builders do not need them, and whoever else wants to build for themselves should also sign for themselves anyway.

**Comment by CompeyDev** [comment]
Kawaii, what are your thoughts on pure FOSS vs a non-free builds separately? As in, the Play Store builds can depend on the billing API, but I would much rather prefer a free build on F-Droid without any of that Google Play reliance. 

I believe some projects like [Keyguard](https://github.com/artemchep/Keyguard) do this, where the F-Droid apps have no paywalls whereas Play ones do. I think this makes sense considering the number of F-Droid users are significantly lesser than Play Store users, so I don't think further effort wouldn't practically be worth the amount of subscriptions you would potentially be getting.

**Comment by kawaiiDango** [comment]
There won't be pure foss builds (as in compatible with the official f-droid repo), due to reasons I have previously stated in this thread. However, there will be builds with acrcloud (needed for Scrobble from mic) as the only non-free dependency in the future.

**Comment by shuvashish76** [comment]
@IzzySoft Recently (Feb 2) they've added additional https://github.com/kawaiiDango/pano-scrobbler#exceptions
Isn't it already violation of GPL-3 license?

**Comment by IzzySoft** [comment]
The image part would mean "non-free assets", but

>  you are FORBIDDEN to publish the builds compiled by you

Is clearly a violation – not only of GPL-3, but also of the [The Four Essential Freedoms of Free Software](https://en.wikipedia.org/wiki/The_Free_Software_Definition#The_Four_Essential_Freedoms_of_Free_Software) – more precisely, freedom-3:

> The freedom to distribute copies of your modified versions to others

To not distribute the original app under one's own name is a different thing – but to forbid modified versions to be distributes as their author wishes (given it makes clear it's a copy, and give credits etc) contradicts the definition of FOSS and GPL.

So @kawaiiDango please update those "exceptions" accordingly.

That said, is Pano-Scrobbler currently listed in any F-Droid repo (except for the single-app one provided by the authors themselves)?

**Comment by kawaiiDango** [comment]
The exceptions were taken from https://github.com/RikkaApps/Shizuku I was unaware that this could be a gpl-3 violation.

**Comment by IzzySoft** [comment]
Yeah, the Shizuku **app** (not the library) has the same issue, which was brought up multiple times already at F-Droid and in other places (and lead to e.g. other developers refusing to include the Shizuku API, as that would require the Shizuku app on the device and such mean a dependency on something non-libre).

**Comment by kawaiiDango** [comment]
I had added the exceptions this year after I heard about the WordPress drama.

Anyways, they are gone now.

**Comment by CompeyDev** [comment]
Hey! Still interested in this, I take that it is currently not possible to simply compile it without all the non-free dependencies.

I'm personally not too familiar with the codebase and have some *basic* Android Dev experience, but I'm curious how we can get started on a refactor which makes it possible for us to only depend acrcloud, having a significantly more free build.

**Comment by kawaiiDango** [comment]
There is a "releaseGithub" build variant which does what you want. The other build variant "release" is meant for play store releases and contains the non free dependencies. 

However, the main branch is a WIP rewrite in compose multiplatform. Most of the UI is currently unusable, although it does scrobble. So, there won't be a release build from the new codebase from my side any time soon.

**Comment by IzzySoft** [comment]
> There is a "releaseGithub" build variant which does what you want

So once the main branch is stable again, will you provide an APK for that build variant as well?

**Comment by kawaiiDango** [comment]
Yes. That's the intent.

**Comment by kawaiiDango** [comment]
I'm considering removing Acrcloud and the microphone permission entirely, even for the play store builds, since according to https://github.com/kawaiiDango/pano-scrobbler/issues/488 Shazam can now show notifications even for normal Shazams, not just AutoShazam, which Pano can then read and scrobble.

**Comment by kawaiiDango** [comment]
Ko-fi or buymeacoffee donations can now be used to unlock the premium features, for builds downloaded from Github releases (or the custom f-droid repo)
